packages/toolpad-app/src/runtime/ToolpadApp.tsx

```diff
@@ -145,6 +153,213 @@ function useElmToolpadComponent(elm: appDom.ElementNode): ToolpadComponent {
   return useComponent(componentId);
 }
 
+interface ParseBindingOptions {
+  scopePath?: string;
+}
+
+/**
+ * Turns an object consisting of a nested structure of BindableAttrValues
+ * into a flat array of relative paths associated with their value.
+ * Example:
+ *   { foo: { bar: { type: 'const', value:1 } }, baz: [{ type: 'jsExpression', value: 'quux' }] }
+ *   =>
+ *   [['.foo.bar', { type: 'const', value:1 }],
+ *    ['.baz[0]', { type: 'jsExpression', value: 'quux' }]]
+ */
+function flattenNestedBindables(
+  params?: NestedBindableAttrs,
+  prefix = '',
+): [string, BindableAttrValue<any>][] {
+  if (!params) {
+    return [];
+  }
+  if (Array.isArray(params)) {
+    return params.flatMap((param, i) => {
+      return flattenNestedBindables(param[1], `${prefix}[${i}][1]`);
+    });
+  }
+  // TODO: create a marker in bindables (similar to $$ref) to recognize them automatically
+  // in a nested structure. This would allow us to build deeply nested structures
+  if (typeof params.type === 'string') {
+    return [[prefix, params as BindableAttrValue<any>]];
+  }
+  return Object.entries(params).flatMap(([key, param]) =>
+    flattenNestedBindables(param, `${prefix}.${key}`),
+  );
+}
+
+function parseBinding(
+  bindable: BindableAttrValue<any>,
+  { scopePath }: ParseBindingOptions = {},
+): ParsedBinding | EvaluatedBinding {
+  if (bindable?.type === 'const') {
+    return {
+      scopePath,
+      result: { value: bindable.value },
+    };
+  }
+  if (bindable?.type === 'jsExpression') {
+    return {
+      scopePath,
+      expression: bindable.value,
+    };
+  }
+  return {
+    scopePath,
+    result: { value: undefined },
+  };
+}
+
+function parseBindings(
```

Comment:
I just moved a few functions above in this file to be able to use them anywhere, and replaced the `page: appDom.PageNode` argument with `rootNode: appDom.ElementNode | appDom.PageNode`, but I actually ended up not using this so far, not sure if it's needed...